File: labs/algorithms-live-coding/java/src/main/java/learning/livecoding/java/PaginationMergeExercise.java
package learning.livecoding.java;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public final class PaginationMergeExercise implements Exercise {
    @Override
    public String id() {
        return "pagination-merge";
    }

    @Override
    public String title() {
        return "Merge Sorted Pages with Two Pointers";
    }

    @Override
    public String summary() {
        return "Walk two already-sorted pages with two pointers, take the newer item each time, skip duplicate IDs, and stop at the response limit.";
    }

    record FeedItem(String id, long updatedAt, String source) {
    }

    static List<FeedItem> mergePages(List<FeedItem> firstPage, List<FeedItem> secondPage, int limit) {
        int firstIndex = 0;
        int secondIndex = 0;
        List<FeedItem> merged = new ArrayList<>();
        Set<String> seenIds = new HashSet<>();

        while (merged.size() < limit && (firstIndex < firstPage.size() || secondIndex < secondPage.size())) {
            FeedItem candidate;
            if (firstIndex >= firstPage.size()) {
                candidate = secondPage.get(secondIndex++);
            } else if (secondIndex >= secondPage.size()) {
                candidate = firstPage.get(firstIndex++);
            } else if (firstPage.get(firstIndex).updatedAt() >= secondPage.get(secondIndex).updatedAt()) {
                candidate = firstPage.get(firstIndex++);
            } else {
                candidate = secondPage.get(secondIndex++);
            }

            if (seenIds.add(candidate.id())) {
                merged.add(candidate);
            }
        }

        return merged;
    }

    @Override
    public void run() {
        List<FeedItem> warehousePage = List.of(
            new FeedItem("item-1", 300, "warehouse"),
            new FeedItem("item-3", 250, "warehouse"),
            new FeedItem("item-5", 180, "warehouse")
        );
        List<FeedItem> storefrontPage = List.of(
            new FeedItem("item-2", 280, "storefront"),
            new FeedItem("item-3", 250, "storefront"),
            new FeedItem("item-4", 220, "storefront")
        );

        List<FeedItem> merged = mergePages(warehousePage, storefrontPage, 4);
        ExerciseSupport.expectEquals(
            "merged-feed-order",
            List.of("item-1", "item-2", "item-3", "item-4"),
            merged.stream().map(FeedItem::id).toList()
        );
    }
}

File: labs/algorithms-live-coding/kotlin/src/main/kotlin/learning/livecoding/kotlin/PaginationMergeExercise.kt
package learning.livecoding.kotlin

object PaginationMergeExercise : Exercise {
    override val id = "pagination-merge"
    override val title = "Merge Sorted Pages with Two Pointers"
    override val summary =
        "Walk two already-sorted pages with two pointers, take the newer item each time, skip duplicate IDs, and stop at the response limit."

    data class FeedItem(val id: String, val updatedAt: Long, val source: String)

    fun mergePages(
        firstPage: List<FeedItem>,
        secondPage: List<FeedItem>,
        limit: Int,
    ): List<FeedItem> {
        var firstIndex = 0
        var secondIndex = 0
        val merged = mutableListOf<FeedItem>()
        val seenIds = mutableSetOf<String>()

        while (merged.size < limit && (firstIndex < firstPage.size || secondIndex < secondPage.size)) {
            val candidate = when {
                firstIndex >= firstPage.size -> secondPage[secondIndex++]
                secondIndex >= secondPage.size -> firstPage[firstIndex++]
                firstPage[firstIndex].updatedAt >= secondPage[secondIndex].updatedAt -> firstPage[firstIndex++]
                else -> secondPage[secondIndex++]
            }

            if (seenIds.add(candidate.id)) {
                merged += candidate
            }
        }

        return merged
    }

    override fun run() {
        val warehousePage = listOf(
            FeedItem("item-1", updatedAt = 300, source = "warehouse"),
            FeedItem("item-3", updatedAt = 250, source = "warehouse"),
            FeedItem("item-5", updatedAt = 180, source = "warehouse"),
        )
        val storefrontPage = listOf(
            FeedItem("item-2", updatedAt = 280, source = "storefront"),
            FeedItem("item-3", updatedAt = 250, source = "storefront"),
            FeedItem("item-4", updatedAt = 220, source = "storefront"),
        )

        val merged = mergePages(warehousePage, storefrontPage, limit = 4)
        ExerciseSupport.expectEquals(
            "merged-feed-order",
            listOf("item-1", "item-2", "item-3", "item-4"),
            merged.map { it.id },
        )
    }
}
